File: delivery_partner_app/lib/models/delivery_order.dart
import 'package:hive/hive.dart';

part 'delivery_order.g.dart';

@HiveType(typeId: 2)
class DeliveryOrder extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String userId;

  @HiveField(2)
  final String restaurantId;

  @HiveField(3)
  final String restaurantName;

  @HiveField(4)
  final List<OrderItem> orderItems;

  @HiveField(5)
  final double orderTotal;

  @HiveField(6)
  final double deliveryFee;

  @HiveField(7)
  final double grandTotal;

  @HiveField(8)
  final LocationModel pickupLocation;

  @HiveField(9)
  final LocationModel deliveryLocation;

  @HiveField(10)
  final String orderStatus;

  @HiveField(11)
  final String paymentStatus;

  @HiveField(12)
  final String paymentMethod;

  @HiveField(13)
  final DateTime orderDate;

  @HiveField(14)
  final DateTime? estimatedDeliveryTime;

  @HiveField(15)
  final DateTime? actualDeliveryTime;

  @HiveField(16)
  final String? deliveryPartnerId;

  @HiveField(17)
  final String? specialInstructions;

  @HiveField(18)
  final String customerName;

  @HiveField(19)
  final String customerPhone;

  @HiveField(20)
  final double? distance;

  DeliveryOrder({
    required this.id,
    required this.userId,
    required this.restaurantId,
    required this.restaurantName,
    required this.orderItems,
    required this.orderTotal,
    required this.deliveryFee,
    required this.grandTotal,
    required this.pickupLocation,
    required this.deliveryLocation,
    required this.orderStatus,
    required this.paymentStatus,
    required this.paymentMethod,
    required this.orderDate,
    this.estimatedDeliveryTime,
    this.actualDeliveryTime,
    this.deliveryPartnerId,
    this.specialInstructions,
    required this.customerName,
    required this.customerPhone,
    this.distance,
  });

  factory DeliveryOrder.fromJson(Map<String, dynamic> json) {
    return DeliveryOrder(
      id: json['_id'] ?? json['id'] ?? '',
      userId: json['userId'] ?? '',
      restaurantId: json['restaurantId'] ?? '',
      restaurantName: json['restaurantName'] ?? '',
      orderItems: (json['orderItems'] as List? ?? [])
          .map((item) => OrderItem.fromJson(item))
          .toList(),
      orderTotal: (json['orderTotal'] as num?)?.toDouble() ?? 0.0,
      deliveryFee: (json['deliveryFee'] as num?)?.toDouble() ?? 0.0,
      grandTotal: (json['grandTotal'] as num?)?.toDouble() ?? 0.0,
      pickupLocation: LocationModel.fromJson(json['pickupLocation'] ?? {}),
      deliveryLocation: LocationModel.fromJson(json['deliveryLocation'] ?? {}),
      orderStatus: json['orderStatus'] ?? '',
      paymentStatus: json['paymentStatus'] ?? '',
      paymentMethod: json['paymentMethod'] ?? '',
      orderDate: DateTime.parse(json['orderDate'] ?? DateTime.now().toIso8601String()),
      estimatedDeliveryTime: json['estimatedDeliveryTime'] != null
          ? DateTime.parse(json['estimatedDeliveryTime'])
          : null,
      actualDeliveryTime: json['actualDeliveryTime'] != null
          ? DateTime.parse(json['actualDeliveryTime'])
          : null,
      deliveryPartnerId: json['deliveryPartnerId'],
      specialInstructions: json['specialInstructions'],
      customerName: json['customerName'] ?? '',
      customerPhone: json['customerPhone'] ?? '',
      distance: (json['distance'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      'restaurantId': restaurantId,
      'restaurantName': restaurantName,
      'orderItems': orderItems.map((item) => item.toJson()).toList(),
      'orderTotal': orderTotal,
      'deliveryFee': deliveryFee,
      'grandTotal': grandTotal,
      'pickupLocation': pickupLocation.toJson(),
      'deliveryLocation': deliveryLocation.toJson(),
      'orderStatus': orderStatus,
      'paymentStatus': paymentStatus,
      'paymentMethod': paymentMethod,
      'orderDate': orderDate.toIso8601String(),
      'estimatedDeliveryTime': estimatedDeliveryTime?.toIso8601String(),
      'actualDeliveryTime': actualDeliveryTime?.toIso8601String(),
      'deliveryPartnerId': deliveryPartnerId,
      'specialInstructions': specialInstructions,
      'customerName': customerName,
      'customerPhone': customerPhone,
      'distance': distance,
    };
  }
}

@HiveType(typeId: 3)
class LocationModel extends HiveObject {
  @HiveField(0)
  final double latitude;

  @HiveField(1)
  final double longitude;

  @HiveField(2)
  final String address;

  @HiveField(3)
  final String? landmark;

  LocationModel({
    required this.latitude,
    required this.longitude,
    required this.address,
    this.landmark,
  });

  factory LocationModel.fromJson(Map<String, dynamic> json) {
    return LocationModel(
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
      address: json['address'] ?? '',
      landmark: json['landmark'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
      'address': address,
      'landmark': landmark,
    };
  }
}

@HiveType(typeId: 4)
class OrderItem extends HiveObject {
  @HiveField(0)
  final String foodId;

  @HiveField(1)
  final String foodName;

  @HiveField(2)
  final int quantity;

  @HiveField(3)
  final double price;

  @HiveField(4)
  final List<String> additives;

  @HiveField(5)
  final String? instructions;

  OrderItem({
    required this.foodId,
    required this.foodName,
    required this.quantity,
    required this.price,
    required this.additives,
    this.instructions,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      foodId: json['foodId'] ?? '',
      foodName: json['foodName'] ?? '',
      quantity: json['quantity'] ?? 0,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      additives: List<String>.from(json['additives'] ?? []),
      instructions: json['instructions'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'foodId': foodId,
      'foodName': foodName,
      'quantity': quantity,
      'price': price,
      'additives': additives,
      'instructions': instructions,
    };
  }
}

File: Backend/models/DeliveryOrder.js
import mongoose from 'mongoose';

const orderItemSchema = new mongoose.Schema({
    foodId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Food',
        required: true
    },
    name: {
        type: String,
        required: true
    },
    price: {
        type: Number,
        required: true
    },
    quantity: {
        type: Number,
        required: true,
        min: 1
    },
    specialInstructions: {
        type: String,
        default: ''
    }
});

const deliveryOrderSchema = new mongoose.Schema({
    orderId: {
        type: String,
        required: true,
        unique: true
    },
    customerId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    restaurantId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Restaurant',
        required: true
    },
    deliveryPartnerId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'DeliveryPartner',
        default: null
    },

    // Order Details
    items: [orderItemSchema],
    orderTotal: {
        type: Number,
        required: true
    },
    deliveryFee: {
        type: Number,
        required: true,
        default: 0
    },
    taxes: {
        type: Number,
        default: 0
    },
    discount: {
        type: Number,
        default: 0
    },
    grandTotal: {
        type: Number,
        required: true
    },

    // Addresses
    restaurantAddress: {
        street: String,
        city: String,
        state: String,
        zipCode: String,
        coordinates: {
            type: [Number], // [longitude, latitude]
            required: true
        }
    },
    deliveryAddress: {
        street: String,
        city: String,
        state: String,
        zipCode: String,
        coordinates: {
            type: [Number], // [longitude, latitude]
            required: true
        },
        apartmentNumber: String,
        landmark: String,
        instructions: String
    },

    // Status and Timing
    orderStatus: {
        type: String,
        enum: [
            'pending',           // Order placed by customer
            'confirmed',         // Restaurant confirmed
            'preparing',         // Restaurant is preparing
            'ready_for_pickup',  // Ready for delivery partner
            'assigned',          // Assigned to delivery partner
            'picked_up',         // Picked up by delivery partner
            'out_for_delivery',  // On the way to customer
            'delivered',         // Successfully delivered
            'cancelled',         // Order cancelled
            'refunded'          // Order refunded
        ],
        default: 'pending'
    },
    deliveryStatus: {
        type: String,
        enum: [
            'not_assigned',
            'assigned',
            'heading_to_restaurant',
            'at_restaurant',
            'picked_up',
            'heading_to_customer',
            'delivered'
        ],
        default: 'not_assigned'
    },

    // Payment
    paymentMethod: {
        type: String,
        enum: ['cash', 'card', 'upi', 'wallet'],
        required: true
    },
    paymentStatus: {
        type: String,
        enum: ['pending', 'paid', 'failed', 'refunded'],
        default: 'pending'
    },
    isPaid: {
        type: Boolean,
        default: false
    },

    // Delivery Details
    estimatedDeliveryTime: {
        type: Date
    },
    actualDeliveryTime: {
        type: Date
    },
    deliveryInstructions: {
        type: String
    },
    contactlessDelivery: {
        type: Boolean,
        default: false
    },

    // Tracking
    timeline: [{
        status: String,
        timestamp: {
            type: Date,
            default: Date.now
        },
        updatedBy: {
            type: String,
            enum: ['customer', 'restaurant', 'delivery_partner', 'system']
        },
        notes: String
    }],

    // Communication
    customerPhone: String,
    restaurantPhone: String,

    // Delivery Partner Earnings
    deliveryEarnings: {
        baseFee: { type: Number, default: 0 },
        distanceBonus: { type: Number, default: 0 },
        timeBonus: { type: Number, default: 0 },
        tip: { type: Number, default: 0 },
        total: { type: Number, default: 0 }
    },

    // Rating and Feedback
    customerRating: {
        rating: { type: Number, min: 1, max: 5 },
        review: String,
        timestamp: Date
    },

    // Distance and Time
    distanceFromRestaurant: Number, // in meters
    distanceToCustomer: Number, // in meters
    estimatedPickupTime: Date,
    actualPickupTime: Date,

    // Special Flags
    isUrgent: {
        type: Boolean,
        default: false
    },
    isReturned: {
        type: Boolean,
        default: false
    },

    // Metadata
    createdAt: {
        type: Date,
        default: Date.now
    },
    updatedAt: {
        type: Date,
        default: Date.now
    }
}, {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
});

// Indexes
deliveryOrderSchema.index({ orderId: 1 });
deliveryOrderSchema.index({ customerId: 1 });
deliveryOrderSchema.index({ restaurantId: 1 });
deliveryOrderSchema.index({ deliveryPartnerId: 1 });
deliveryOrderSchema.index({ orderStatus: 1 });
deliveryOrderSchema.index({ deliveryStatus: 1 });
deliveryOrderSchema.index({ createdAt: -1 });
deliveryOrderSchema.index({ 'restaurantAddress.coordinates': '2dsphere' });
deliveryOrderSchema.index({ 'deliveryAddress.coordinates': '2dsphere' });

// Virtual for delivery duration
deliveryOrderSchema.virtual('deliveryDuration').get(function () {
    if (this.actualDeliveryTime && this.actualPickupTime) {
        return Math.round((this.actualDeliveryTime - this.actualPickupTime) / (1000 * 60)); // in minutes
    }
    return null;
});

// Virtual for total duration
deliveryOrderSchema.virtual('totalDuration').get(function () {
    if (this.actualDeliveryTime && this.createdAt) {
        return Math.round((this.actualDeliveryTime - this.createdAt) / (1000 * 60)); // in minutes
    }
    return null;
});

// Method to add timeline entry
deliveryOrderSchema.methods.addToTimeline = function (status, updatedBy, notes = '') {
    this.timeline.push({
        status,
        updatedBy,
        notes,
        timestamp: new Date()
    });
    return this;
};

// Method to update order status
deliveryOrderSchema.methods.updateStatus = async function (newStatus, updatedBy, notes = '') {
    this.orderStatus = newStatus;
    this.addToTimeline(newStatus, updatedBy, notes);
    this.updatedAt = new Date();

    // Update delivery status based on order status
    if (newStatus === 'assigned') {
        this.deliveryStatus = 'assigned';
    } else if (newStatus === 'picked_up') {
        this.deliveryStatus = 'picked_up';
        this.actualPickupTime = new Date();
    } else if (newStatus === 'out_for_delivery') {
        this.deliveryStatus = 'heading_to_customer';
    } else if (newStatus === 'delivered') {
        this.deliveryStatus = 'delivered';
        this.actualDeliveryTime = new Date();
    }

    return await this.save();
};

// Static method to find nearby orders for delivery partners
deliveryOrderSchema.statics.findAvailableOrders = function (longitude, latitude, maxDistance = 10000) {
    return this.find({
        orderStatus: 'ready_for_pickup',
        deliveryPartnerId: null,
        'restaurantAddress.coordinates': {
            $near: {
                $geometry: {
                    type: 'Point',
                    coordinates: [longitude, latitude]
                },
                $maxDistance: maxDistance
            }
        }
    })
        .populate('restaurantId', 'name phone address')
        .populate('customerId', 'name phone')
        .sort({ createdAt: 1 });
};

// Static method to find orders by delivery partner
deliveryOrderSchema.statics.findByDeliveryPartner = function (deliveryPartnerId, status = null) {
    const query = { deliveryPartnerId };
    if (status) {
        query.orderStatus = status;
    }

    return this.find(query)
        .populate('restaurantId', 'name phone address')
        .populate('customerId', 'name phone')
        .sort({ createdAt: -1 });
};

const DeliveryOrder = mongoose.model('DeliveryOrder', deliveryOrderSchema);

export default DeliveryOrder;
